### DISCUSSION TITLE
Hosting on specified domains breaks OpenAPI docs

### DISCUSSION THREAD

**Discussion by tsunyoku** [fastapi/fastapi#8464] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, APIRouter

app = FastAPI()
router = APIRouter()

@router.get("/")
def read_root():
    return {"Hello": "World"}

app.host("test.localhost", router)
```

### Description

When using host to limit routers to domains, accessing the /docs page on the specified domains always fails to find the routes linked with that router.

The expected behaviour would be each domain to list all the routes on their respective routers, so in the example given I would expect `test.localhost/docs` to show the index route, but nothing is shown.

### Operating System

Linux, Windows

### FastAPI Version

0.75.0

### Python Version

3.9.10

**Comment by accelleon** [comment]
Starlette discourages the use of `Starlette.host()`
```python
        """
        We no longer document this API, and its usage is discouraged.
        Instead you should use the following approach:
        routes = [
            Host(path, ...),
            ...
        ]
        app = Starlette(routes=routes)
        """
```

Anywho the issue here is that `.host()` is a Starlette function, not a FastAPI one. By using `.host()` instead of `.include_router()` you're feeding the route to Starlette and not FastAPI, and hence is bypassing all of the fancy FastAPI specific stuff such as the OpenAPI doc generation for your router.

**Comment by tsunyoku** [comment]
is there any way within FastAPI to achieve the same domain-limiting behaviour that `.host()` provides then?

**Comment by accelleon** [comment]
Since its provided by Starlette, it can be done. Easily? I'm not sure. Perhaps something like subclassing `APIRoute` and handling the addition of the server override to the generated OpenAPI docs would do it.

Personally, I'd be using a proxy like traefik or nginx. I admit I'm not exactly sure in what circumstances you'd want this behavior other than you're hosting a frontend and backend on the same server.

**Comment by tsunyoku** [comment]
Nginx works if we only use one router on a server, but in my case the server is handling multiple routers on different subdomains for a game network. It handles different tasks over different subdomains which we cannot control and some have clashing endpoints, hence the need for domain-routing in FastAPI aswell as nginx.

**Comment by accelleon** [comment]
Someone whose dug deeper than I into FastAPI's internals could probably point you in the right direction to get it done in FastAPI with OpenAPI documentation.

As for dealing with Nginx, provided your specific use case does allow this: (and i'm not gloriously misunderstanding)
```python
from fastapi import FastAPI, APIRouter

app = FastAPI()
route1 = APIRouter(prefix='/route1/')
route2 = APIRouter(prefix='/route2/')

@route1.get(path="/")
def get1():
    return {"Hello": "World"}

@route2.get(path="/")
def get2():
    return {"Hello": "World2"}

app.include_router(route1)
app.include_router(route2)
```
```
server {
  listen 80;
  server_name route1.localhost;
  location / {
    proxy_pass http://127.0.0.1:8000/route1/;
  }
}

server {
  listen 80;
  server_name route2.localhost;
  location / {
    proxy_pass http://127.0.0.1:8000/route2/;
  }
}
```
Granted your docs would still be at just `localhost/docs` and they'd be for all routes.
The only way I see to split the docs if that's what you're after would be to host multiple ASGI server instances against multiple different apps or heavily modify FastAPI.

**Comment by YuriiMotov** [comment, marked as answer]
Seems that [sub-applications](https://fastapi.tiangolo.com/advanced/sub-applications/) + proxy (as described [in this comment](https://github.com/fastapi/fastapi/discussions/8464#discussioncomment-5152322)) should help with this.

Or even without sub-applications (if there is no need for you to split the docs)
